app/pages/learn/01_tutorial/08_more-resources/02_debugging.md

```diff
@@ -0,0 +1,136 @@
+
+A debugger is a tool in your IDE (integrated development environment) that lets you see the values of different variables at different points in the program. It's like a really powerful magnifying glass. While a few details vary between Eclipse/IntelliJ/NetBeans/VS Code, the concepts are the same.
+
+<a id="println">&nbsp;</a>
+## Why not Println
+
+When first learning how to code, we often write code like this to see what is going on:
+
+```java
+System.out.println(numCats);
+```
+
+There's nothing wrong with using println (as long as you don't commit it). However, it can quickly get unmanagable in a more complicated program especally if you have multiple things to keep track of or a lot of loops. It can also be hard to find when there is a lot of logging in the application. (I used the stars and my initials to mitigate that, but still can get past the point where println is useful.)
+
+```java
+System.out.println("*** JB i=%s numCats=%s numDogs=%s".formatted(i, numCats, numDogs));
+```
+
+Even if println() does meet your needs at the moment, it won't forever. Learning how to use a debugger helps avoid the problem that you use println just because it is all you know.
+
+<a id="unit">&nbsp;</a>
+## Why not unit testing
+
+Nothing is wrong with unit testing. Unit tests are great. They document expected behavior. They tell you if an unexpected value is returned. They help you understand the behavior of the code. And sometimes they can even give you big clues about what is wrong with the code. However, they don't tell you what is happening inside the broken code when it isn't returning the right value. For that, you use a debugger with the unit test to see what is going on inside the method.
+
+```java
+@Test
+void magic() {
+     assertEquals(42, target.magic(), "magic number incorrect");
+}
+```
+
+<a id="breakpoint">&nbsp;</a>
+## What is a breakpoint?
+
+I've (improperly) implemented the magic method. It's supposed to multiply six times seven and get 42. However, that's not what happens. After (not) much investigation, I am baffled and realize I want to know what the values of part1 and part2 are on line 5.
+
+That's what a breakpoint is for. It lets me tell the debugger to pause the program there and let me poke around.
+
+```java
+public class Answer {
+    public  int magic() {
+        int part1 = 3 + 2;  // BUG!
+        int part2 = 7;
+        return part1 * part2;
+    }
+}
+```
+
+When the debugger stops on line 5, I see that part2 is the seven that I expected. However part1 is five,, not six. I found the bug! Thanks to the debugger allowing me to set a breakpoint.
```

Comment:
Add \` \` around the references to variables `part1` and `part2` in the paragraph text